rfcs/0003-mutations.md

```diff
@@ -0,0 +1,280 @@
+})
+
+const link = split(
+  ({ query }) => {
+    const node = getMainDefinition(query);
+    return node.kind === "OperationDefinition" &&
+           node.operation === "mutation"
+  },
+  mutationLink,
+  queryLink
+);
+
+// Create Apollo Client
+const client = new ApolloClient({
+  link,
+  cache: new InMemoryCache()
+})
+
+const CREATE_ENTITY = gql`
+  mutation createEntity($options: MyEntityOptions) {
+    createEntity(options: $options) {
+      id
+      name
+      value
+    }
+  }
+`
+
+// state === resolver's state
+const [exec, { loading, state }] = useMutation(
+  CREATE_ENTITY,
+  {
+    client,
+    variables: {
+      options: { name: "...", value: 5 }
+    }
+  }
+)
+
+// We can also utilize an optimistic response
+const [exec, { loading, state }] = useMutation(
+  CREATE_ENTITY,
+  {
+    optimisticResponse: {
+      myEntity: {
+        id: "...",
+        name: "...",
+        value: 5,
+      }
+    },
+    update(proxy, { data }) {
+      // result = data.myEntity
+    },
+    onError(error) {
+      ...
+    },
+    variables: {
+      options: { name: "...", value: 5 }
+    }
+  }
+)
+```
+
+## Compatibility
+
+No breaking changes will be introduced, as mutations are an optional add-on to a subgraph.
+
+## Drawbacks and Risks
+
+I have some thoughts but they are rather verbose and tangential. Would love some feedback on this from others first.
+
+## Alternatives
+
+The existing alternative that protocol developers are creating for dApp developers has been described above.
+
+## Open Questions
+
+- **Should the resolvers module be ES5 compliant?**  
+  We've been operating under this assumption while developing the prototype. We have since scrapped this requirement as it has proven nearly impossible to successfully transpile our own source, along with all our dependencies, into a single monolithic module. If anyone has experience doing this I would love chat!
+- **How should the dApp configure the resolvers module?**  
+  The dApp knows best how to: connect to the various web3 networks, handle key signature requests, and all other user / dApp specific things. We need a way for the dApp to configure the resolvers in a specific way given the resolver's requirements (IPFS provider, Web3 provider, etc).
+- **What paradigm should the resolver state follow?**  
+  One option is to have the resolver's call into a single interface that modifys the backing data. Whenever this data is modified, the entirety of it is passed to the dApp. The downside here is that the dApp doesn't know what has changed within the data, and is forced to represent it in its entirety in order to not miss anything.  
+
+  Another option is to implement something similar to Redux, where the resolvers fire off events with corresponding payloads of data. These events map to reducers, which take in this payload of data and decide what to do with it. The dApp could implement these reducers, and choose how it would want to react to the various events.
```

Comment:
There are a couple things I'm missing in the RFC so far:

1. What exactly does `createMutations` return? General, there's little in terms of a specification included aside from some code snippets.

2. What exactly does the mutation state API look like? When it comes to the paradigm, I don't think we're limited to a single one. To me, the ideal API would include both "static" state information as well as progress information like a progress percentage or a sequence of events. The big question and one that I was hoping we'd already be closer to is: what does this API look like? Here's a strawman:
    ```js
    type MutationEvent = TransactionEvent | ...
    type TransactionEvent = TransactionStartedEvent | TransactionCanceledEvent | TransactionConfirmedEvent | ...

    interface TransactionStartedEvent {
      transaction: Transaction
      ...
    }

    interface Transaction {
      id: string
      title: string
      ...
    }

    interface MutationState {
      progress: Number, // Between 0.0 and 1.0
      events: MutationEvent[],

      startTransaction({ id: string, title: string, ... })
      cancelTransaction(id: string, ...)
      confirmTransaction(id: string, ...)
      ...
    }
    ```
  what I'm not sure personally is what a good abstraction for transactions is across different chains is. I'm not overly familiar with meta transactions etc. so that makes it hard for me to come up with something on the spot. But this is what we need to think about — unless we l basically make `Transaction` an opaque structure with just an `id`, a human-readable `title`, perhaps a `link` to visit for details and an arbitrary `payload` or something.

Reply:
Note: In the above strawman, the `progress` and `events` should be exposed to the user of `useMutation` but the `startTransaction` etc. functions should not.

Reply:
1. I've added type definitions for all API usages present within the example. Let me know if they're too verbose or if they should be detailed elsewhere in the RFC.
2. I've added the mutation state API to the RFC.